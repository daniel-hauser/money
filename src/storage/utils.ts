import { parseISO, roundToNearestMinutes } from "date-fns";
import type { CompanyTypes } from "israeli-bank-scrapers";
import type { Transaction } from "israeli-bank-scrapers/lib/transactions";

export function transactionHash(
  tx: Transaction,
  companyId: CompanyTypes,
  accountNumber: string,
) {
  const date = roundToNearestMinutes(parseISO(tx.date)).toISOString();
  const parts = [
    date,
    tx.chargedAmount,
    tx.description,
    tx.memo,
    companyId,
    accountNumber,
  ];

  return parts.map((p) => String(p ?? "")).join("_");
}
