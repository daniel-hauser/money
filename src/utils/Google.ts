import { GoogleSpreadsheet } from "google-spreadsheet";
import { GoogleAuth } from "google-auth-library";
import { createLogger } from "../utils/logger.js";

const logger = createLogger("utils/Google");

export async function getGoogleSheet(
  serviceAccountEmail: string,
  serviceAccountPrivateKey: string,
  sheetId: string,
) {
  try {
    const auth = new GoogleAuth({
      scopes: ["https://www.googleapis.com/auth/spreadsheets"],
      credentials: {
        client_email: serviceAccountEmail,
        private_key: serviceAccountPrivateKey,
      },
    });

    const doc = new GoogleSpreadsheet(sheetId, auth);
    await doc.loadInfo();
    return doc;
  } catch (error) {
    logger(`Error fetching google sheet: ${error}`);
    throw error;
  }
}

export async function exportGSheetToCSV(
  googleCredentials: any,
  sheetId: string | undefined,
  worksheetName: string | undefined,
): Promise<string> {
  if (
    !googleCredentials.client_email ||
    !googleCredentials.private_key ||
    !sheetId ||
    !worksheetName
  ) {
    throw new Error("All google arguments must be passed");
  }
  try {
    const doc = await getGoogleSheet(
      googleCredentials.client_email,
      googleCredentials.private_key,
      sheetId,
    );
    const sheet = doc.sheetsByTitle[worksheetName];
    if (!sheet) {
      throw new Error(`Sheet ${worksheetName} not found`);
    }
    const arrayBuffer = await sheet.downloadAsCSV();
    const csvData = arrayBufferToString(arrayBuffer);
    return csvData;
  } catch (error) {
    logger(`Error exporting google sheet to CSV ${error}`);
    throw error;
  }
}

function arrayBufferToString(buffer: ArrayBuffer): string {
  return new TextDecoder("utf-8").decode(new Uint8Array(buffer));
}
