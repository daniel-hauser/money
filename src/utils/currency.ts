import { TransactionRow } from "../types.js";

const symbolToCurrency = {
  $: "USD",
  "€": "EUR",
  "₪": "ILS",
};

export function normalizeCurrency(currency: string | undefined) {
  if (!currency) return undefined;
  return (symbolToCurrency[currency] ?? currency).toUpperCase();
}

const ILS_CURRENCY_LABELS = ["₪", "ILS"];

/**
 * Foreign currencies resolve to ILS when settled after changing from pending status
 * @param tx Transaction checked
 * @returns True if tx resolved in ILS currency
 */
export function checkIlsCharge(tx: TransactionRow): boolean {
  return (
    (tx.chargedCurrency != undefined &&
      ILS_CURRENCY_LABELS.includes(tx.chargedCurrency.trim())) ||
    (tx.originalCurrency != undefined &&
      ILS_CURRENCY_LABELS.includes(tx.originalCurrency.trim()))
  );
}
