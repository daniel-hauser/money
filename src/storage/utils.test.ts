import {
  Transaction,
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions";
import { transactionHash } from "./utils";
import { CompanyTypes } from "israeli-bank-scrapers";

describe("transactionHash", () => {
  const transaction1: Transaction = {
    date: "2021-01-01",
    processedDate: "2021-01-01",
    chargedAmount: 1,
    description: "desc 1",
    memo: "memo 1",
    originalAmount: 1,
    originalCurrency: "ILS",
    status: TransactionStatuses.Completed,
    type: TransactionTypes.Normal,
  };

  const transaction2: Transaction = {
    date: "2021-01-01",
    processedDate: "2021-01-01",
    chargedAmount: 1,
    description: "desc 2",
    memo: "memo 2",
    originalAmount: 1,
    originalCurrency: "ILS",
    status: TransactionStatuses.Completed,
    type: TransactionTypes.Normal,
  };

  it("should return the same hash for the same transaction", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const hash1 = transactionHash(transaction1, companyId, accountNumber);
    const hash2 = transactionHash(transaction1, companyId, accountNumber);

    expect(hash1).toBe(hash2);
  });

  it("should return different hashes for different transactions", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const hash1 = transactionHash(transaction1, companyId, accountNumber);
    const hash2 = transactionHash(transaction2, companyId, accountNumber);

    expect(hash1).not.toBe(hash2);
  });

  it("should return different hashes for different company ids", () => {
    const accountNumber = "123";
    const hash1 = transactionHash(
      transaction1,
      CompanyTypes.leumi,
      accountNumber
    );

    const hash2 = transactionHash(
      transaction1,
      CompanyTypes.isracard,
      accountNumber
    );

    expect(hash1).not.toBe(hash2);
  });

  it("should return different hashes for different account numbers", () => {
    const companyId = CompanyTypes.hapoalim;
    const hash1 = transactionHash(transaction1, companyId, "123");
    const hash2 = transactionHash(transaction1, companyId, "456");

    expect(hash1).not.toBe(hash2);
  });

  it("Issue #132: hash has no nullish string", () => {
    const transactionWithUndefined = {
      ...transaction1,
      chargedAmount: undefined,
      description: undefined,
      memo: undefined,
    };

    const transactionWithNull = {
      ...transaction1,
      chargedAmount: null,
      description: null,
      memo: null,
    };

    for (const transaction of [transactionWithNull, transactionWithUndefined]) {
      const hash = transactionHash(
        transaction as any,
        CompanyTypes.leumi,
        "123"
      );

      expect(hash).not.toContain("null");
      expect(hash).not.toContain("undefined");
    }
  });
});
