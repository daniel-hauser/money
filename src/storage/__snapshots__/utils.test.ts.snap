// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactionUniqueId should return the same unique id for the same transaction 1`] = `"hapoalim_123_2021-01-01_1000_identifier1"`;
